lib/src/git_backend.rs

```diff
@@ -518,19 +528,32 @@ fn commit_from_git_without_root_parent(
         .try_to_commit_ref()
         .map_err(|err| to_read_object_err(err, id))?;
 
-    // We reverse the bits of the commit id to create the change id. We don't want
-    // to use the first bytes unmodified because then it would be ambiguous
-    // if a given hash prefix refers to the commit id or the change id. It
-    // would have been enough to pick the last 16 bytes instead of the
-    // leading 16 bytes to address that. We also reverse the bits to make it less
-    // likely that users depend on any relationship between the two ids.
-    let change_id = ChangeId::new(
-        id.as_bytes()[4..HASH_LENGTH]
-            .iter()
-            .rev()
-            .map(|b| b.reverse_bits())
-            .collect(),
-    );
+    // If the git header has a change-id field, we attempt to convert that to a
+    // valid JJ Change Id
+    let change_id = commit
+        .extra_headers()
+        .find("change-id")
+        .iter()
+        .filter_map(|reverse_hex| to_forward_hex(reverse_hex.to_string().as_str()))
+        .filter_map(|change_id_hex| ChangeId::try_from_hex(change_id_hex.as_str()).ok())
```

Comment:
nit: Instead of `.iter().filter_map().filter_map()`, does `.and_then().and_then()` work?